### DISCUSSION TITLE
null assertion operator https://kotlinlang.org/docs/null-safety.html#the-operator

### DISCUSSION THREAD

**Discussion by jasekiw** [dotnet/csharplang#7951] [Language Ideas]
I have seen similar issues around this but all get derailed by irrelevant details. Other languages have this operator and do it well. Rust calls it Unwrap, kotlin calls it the null assertion operator. https://kotlinlang.org/docs/null-safety.html#the-operator

Right now our best option is an extensions method which creates a lot of extra typing. 

```csharp
public static T ThrowIfNull<T>([NotNull] this T? source)
{
    if (source == null)
    {
        throw new ArgumentNullException();
    }
    return source;
}
```

```csharp
MyThingClass? myThing = ...;
```

```csharp
return myThing.ThrowIfNull().SomeProperty.ThrowIfNull().SomeOtherProperty.ThrowIfNull()
```

To help with verbosity I could do this instead:

```csharp
public static T N<T>([NotNull] this T? source)
{
    if (source == null)
    {
        throw new ArgumentNullException();
    }
    return source;
}
```

```csharp
return myThing.N().SomeProperty.N().SomeOtherProperty.N()
```

But this would be a lot better if it was just supported by the language. I think we can agree that getting in the weeds about the syntax is just caused head ache. At this point just do what other languages are already doing and proved to be useful like the !! operator from kotlin.

```csharp
return myThing!!.SomeProperty!!.SomeOtherProperty!!
```

The case shown above is very common if using nullable navigation properties in ef core. 

As for people wanting to use !! for method parameter not null checking, we already have things like https://github.com/Singulink/RuntimeNullables  that can handle this for you automatically. Something like this could be incorporated into the language for better support under a runtime null check flag like how nullable reference types are enabled now.

**Comment by HaloFour** [comment]
See: #6034

**Comment by colejohnson66** [comment]
```cs
return myThing?.SomeProperty?.SomeOtherProperty ?? throw new ArgumentNullException("myThing");
```

Or use the "dammit operator":
```cs
return myThing!.SomeProperty!.SomeOtherProperty;
```

**Reply by jasekiw** [reply]
This is not equivalent to what was described above. dammit operator has no runtime effect, and the ?? throw new ArgumentNullException does not specify which value was null.
